test_conformance/api/negative_device.cpp

```diff
@@ -0,0 +1,509 @@
+    {
+        printf("Device does not support creating subdevices... Skipping\n");
+        return TEST_SKIPPED_ITSELF;
+    }
+
+    cl_device_partition_property properties[3] = {};
+    cl_uint max_compute_units =
+        get_uint_device_info(deviceID, CL_DEVICE_MAX_COMPUTE_UNITS);
+    cl_uint max_sub_devices =
+        get_uint_device_info(deviceID, CL_DEVICE_PARTITION_MAX_SUB_DEVICES);
+    if (supported_properties & SupportedPartitionSchemes::Equally)
+    {
+        properties[0] = CL_DEVICE_PARTITION_EQUALLY;
+        properties[1] = max_compute_units;
+    }
+    else if (supported_properties & SupportedPartitionSchemes::Counts)
+    {
+        properties[0] = CL_INVALID_DEVICE_PARTITION_COUNT;
+        properties[1] = max_sub_devices;
+    }
+    else
+    {
+        properties[0] = CL_DEVICE_PARTITION_BY_AFFINITY_DOMAIN;
+        properties[1] = CL_DEVICE_AFFINITY_DOMAIN_NEXT_PARTITIONABLE;
+    }
+
+    properties[2] = 0;
+    cl_device_id *out_devices = nullptr;
+
+    cl_int err(CL_SUCCESS);
+    for (auto invalid_device : get_invalid_objects<cl_device_id>(deviceID))
+    {
+        err = clCreateSubDevices(invalid_device, properties, 1, out_devices,
+                                 nullptr);
+        test_failure_error_ret(err, CL_INVALID_DEVICE,
+                               "clCreateSubDevices should return "
+                               "CL_INVALID_DEVICE when: \"in_device "
+                               "is not a valid device\"",
+                               TEST_FAIL);
+    }
+
+    err = clCreateSubDevices(deviceID, nullptr, 1, out_devices, nullptr);
+    test_failure_error_ret(
+        err, CL_INVALID_VALUE,
+        "clCreateSubDevices should return CL_INVALID_VALUE when: \"values "
+        "specified in properties are not valid\" using a nullptr",
+        TEST_FAIL);
+
+    err = clCreateSubDevices(deviceID, properties, 0, out_devices, nullptr);
+    test_failure_error_ret(
+        err, CL_INVALID_VALUE,
+        "clCreateSubDevices should return CL_INVALID_VALUE when: \"out_devices "
+        "is not NULL and num_devices is less than the number of sub-devices "
+        "created by the partition scheme\"",
+        TEST_FAIL);
+
+    if (supported_properties != SupportedPartitionSchemes::All_Schemes)
+    {
+        std::vector<cl_device_partition_property> invalid_properties =
+            get_invalid_properties(supported_properties
+                                   ^ SupportedPartitionSchemes::All_Schemes);
+        err = clCreateSubDevices(deviceID, invalid_properties.data(), 1,
+                                 out_devices, nullptr);
+        test_failure_error_ret(
+            err, CL_INVALID_VALUE,
+            "clCreateSubDevices should return CL_INVALID_VALUE when: \"values "
+            "specified in properties are valid but not supported by the "
+            "device\"",
+            TEST_FAIL);
+    }
+
+    err = clCreateSubDevices(deviceID, properties, properties[1], out_devices,
```

Comment:
Please check that this is correct also - it's passing `properties[1]` as `num_devices`?

Reply:
This is a difficult one as the spec isn't clear on how to ensure that CL_DEVICE_PARTITION_FAILED is returned.
Would appreciate some suggestions on this.

Reply:
Yes, I agree this is a little tricky.

I interpret the spec as saying that `CL_DEVICE_PARTITION_FAILED` will be returned when all parameters are valid but no partitioning will occur.  There are multiple ways to do this, for example: partitioning equally into the same number of compute units as the input device, partitioning by counts into the same number of compute units as the input device, or partitioning by affinity domain when there are no more affinity domains to partition by.

One possible gotcha with this case is that (I think) `properties[1]` (and hence `num_devices`) is non-zero, but `out_devices` is `nullptr`.  I don't see anything prohibiting this in the spec, but it's admittedly a little weird.  Is this valid?  What should happen in this case?